### ISSUE TITLE
DBusException: Method "OpenSession" with signature "ss" on interface "org.freedesktop.Secret.Service" doesn't exist

### ISSUE THREAD

**Issue by jaraco** [jaraco/keyring#65]
Sample code triggering the error:

```
import keyring
keyring.get_password('test', 'test')
```

gives exception 

```
DBusException: org.freedesktop.DBus.Error.UnknownMethod: Method "OpenSession" with signature "ss" on interface "org.freedesktop.Secret.Service" doesn't exist
```

I'm on Archlinux 64bit and gnome-keyring-daemon is running (version 3.4.1).
You may be interested at this bug report: https://answers.launchpad.net/polly/+question/195167

---
- Bitbucket: https://bitbucket.org/kang/python-keyring-lib/issue/65
- Originally reported by: Stéphane Travostino
- Originally created at: 2012-06-11T18:09:06.134

**Comment by jaraco** [comment]
Comment 6 in the linked question says the following:

```
It seems that this is a regression in dbus-python 1.1, and it doesn't convert string to variant automatically (OpenSession signature is "sv").

Changing

    _, session = service_iface.OpenSession("plain", "")

to

    _, session = service_iface.OpenSession("plain", dbus.String("", variant_level=1))

in /usr/lib64/python2.7/site-packages/keyring/backend.py should help here.
```

Patching keyring to address a python-dbus bug doesn't seem like the right thing to do here. I'm not sure what the implications would be. I'm not sure how to detect also when the patch would no longer be necessary.

---

Original comment by: Jason R. Coombs

**Comment by jaraco** [comment]
I'm going to close this as wontfix, at least until such a time that someone reports that having compatibility for dbus-python 1.1 (and working around the bug) has practical value for some project or deployment. Otherwise, let's just wait for the fix in the upstream library.

---

Original comment by: Jason R. Coombs

**Comment by jaraco** [comment]
The dbus-python folks looked at this and decided there was no bug in dbus-python.  See this discussion:  https://bugs.freedesktop.org/show_bug.cgi?id=50370

I checked the DBus interface as described in the discussion, and what I saw _might_ (I'm not familiar with DBus or any of these libraries) support your conclusion that it's a dbus-python bug.  If so, you might want to reopen that bug report, because otherwise nobody is working on it.

Here's how I checked the interface:

```
service_obj = session_bus.get_object('org.freedesktop.secrets', '/org/freedesktop/secrets')
introspectable = dbus.Interface(service_obj, 'org.freedesktop.DBus.Introspectable')
print introspectable.Introspect()
```

Here's the relevant snippet of the output:

```
<interface name="org.freedesktop.Secrets.Service">

    <property name="Collections" type="ao" access="read"/>

    <method name="OpenSession">
        <arg name="algorithm" type="s" direction="in"/>
        <arg name="input" type="v" direction="in"/>
        <arg name="output" type="v" direction="out"/>
        <arg name="result" type="o" direction="out"/>
    </method>
```

I got exactly the same error message as the original reporter.  (I'm trying to start the everpad provider daemon under Debian Wheezy, for what it's worth.)

```
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.UnknownMethod: Method "OpenSession" with signature "ss" on interface "org.freedesktop.Secret.Service" doesn't exist
```

HTH.

---

Original comment by: David Huebel

**Comment by jaraco** [comment]
We should clarify this issue, otherwise this backend becomes useless. There's something weird actually going here : it seems it should be converting the string into a variant anyway (per https://bugs.freedesktop.org/show_bug.cgi?id=50370), and introspecting the interface with dbus-python doesn't return the right methods...

I propose to let this bug "on hold", at least until we really know where the bug is.

---

Original comment by: Jonathan Ballet

**Comment by jaraco** [comment]
My DBus knowledge sucks, @davidhuebel are you interested to have a look on it?

---

Original comment by: Jonathan Ballet

**Comment by jaraco** [comment]
Waiting to know where the problem is. For the time being, the "Secret Service" backend doesn't work, at least with Gnome Keyring (which might be the only one implementing it?)

---

Original comment by: Jonathan Ballet

**Comment by jaraco** [comment]
Finally checked back and decided to have a go at tracking this down.  It turns out the bug is caused by incorrect introspection data returned by gnome-keyring.  dbus-python refuses to dispatch the call because the introspection data provided by gnome-keyring says the method does not exist.  The bug was introduced in gnome-keyring 3.4 and fixed in this commit:  http://git.gnome.org/browse/gnome-keyring/commit/?id=5dccbe88eb94eea9934e2b7c83e818bd21be4fd2

It looks like it should be fixed in gnome-keyring 3.5, but I have not verified this.

---

Original comment by: David Huebel

**Comment by jaraco** [comment]
Sounds like there's not much for keyring to do here. If so, please revive the conversation, preferably with a proposal to fix the issue. If what you supply is a temporary workaround for a bug in another system, please be sure to indicate when that workaround can be safely retired (e.g. "# remove this code when support for Gnome-keyring 3.4 is no longer required").
